### ISSUE TITLE
Blocks should not be federated

### ISSUE THREAD

**Issue by PleromaUser** [tootsuite/mastodon#11503]
### Expected behaviour

Blocks are not sent to remote servers, following ActivityPub spec and preventing blocked users from knowing they're blocked.

### Actual behaviour

Blocks are sent to remote servers.

### Steps to reproduce the problem

1. Make an alt account on another server
2. Block from your main account
3. Your alt account will be able to see that they can no longer see your profile (i.e. they're blocked)

### Specifications

Any

**Comment by PleromaUser** [comment]
(for example, federating blocks allows the following type of abuse to occur: https://git.pleroma.social/pleroma/pleroma/merge_requests/1536)

**Comment by Gargron** [comment]
The specification of the block feature is such that the the blocked user must not be able to see the profile or interact with any posts. If you don't do that, you've got a mute instead of a block. It is not possible to provide functionality that forbids the blocked user from doing something without notifying the blocked user's server.

Before someone chimes in about OCAPs, even an OCAP would not hide the fact that an OCAP was revoked from the blocked user's server.

**Comment by PleromaUser** [comment]
> The specification of the block feature is such that the the blocked user must not be able to see the profile or interact with any posts. If you don't do that, you've got a mute instead of a block. It is not possible to provide functionality that forbids the blocked user from doing something without notifying the blocked user's server.

This is impossible to implement.

**Comment by trwnh** [comment]
i'm sure someone with enough time could figure if they're blocked. but still, "notifying the blocked user's server" is only necessary insofar as actions are processed on the remote server. for not sending a Block to be meaningful, you would have to have side effects rely mainly/purely on your local server. For example, silently discarding their replies or Follow activities, while accepting that they can view public posts.

I would like to see more control handed over to the user, e.g. "do you want to send a Block activity?" which would allow either silent discards or a more explicit notice (if the remote server supports handling received Block activities), but right now it's kind of not as significant a difference until there's an alternative to public posts (e.g. "authorized fetch" or delivering to actor inboxes instead of sharedInbox).

**Comment by PleromaUser** [comment]
Blocks should be handled similarly to Pleroma:

On the blocker's server
* Reject follow requests
* Reject likes, announces, replies, etc to the blocker's posts

On the blockee's server
* Still allow the blocked user to view the blocker's **public** posts (not followers-only)
* Still allow the blocked user to reply

All of this behavior can be maintained without federating blocs

**Comment by trwnh** [comment]
allowing the blocked user to reply is a no-go because replies from a remote user are delivered independently of your local server. you need some signal for the remote server to disallow the reply, or for other remote servers to reject the reply as invalid.

one way, the simplest and what is currently done, is to just deliver the Block, and assume that the other server is a compliant implementation. (if it is not, then the entire domain can be suspended or firewalled.)

the other way, which would require more reworking, is to require some token of validity to be attached to replies. for example, require a signed Accept Follow from the person you are replying to, similar to LD-sigs.

**Comment by PleromaUser** [comment]
Requiring a signed Accepted Follow makes no sense because a user should be able to reply to users they don't follow.

A better solution is to have the parent server list all replies, and any replies not in that list be rejected, but unfortunately, Mastodon only lists self replies, so that won't work.

**Comment by trwnh** [comment]
>a user should be able to reply to users they don't follow.

this is a point of contention. many people would like to prevent replies from being made, or at least not proliferating (because of course it is always possible to host a web page "replying" to any other web page)

>have the parent server list all replies

i agree this would allow the origin server to decide which replies are valid or not. it might slightly increase load because now you have to query the origin server for replies, but it's an option that has been raised before: https://schub.io/blog/2018/02/01/activitypub-one-protocol-to-rule-them-all.html

>As per the ActivityStreams spec, all objects have a `replies` property. So, a more sensible, reliable, and even more AcitivityStream’y way of handling replies would probably be adding the interaction to the replies collection and sending an `update`.

of course the key difference being that this would make posts no longer independent objects, but dependent on the parent context (e.g. less so "replies" and more so "comments"). there's also the issue of addressing to your own followers vs. to the followers of the parent post's author.

**Comment by PleromaUser** [comment]
> this is a point of contention. many people would like to prevent replies from being made, or at least not proliferating (because of course it is always possible to host a web page "replying" to any other web page)

The vast majority wouldn't.  Those that do should post followers-only.

**Comment by trwnh** [comment]
i thing the majority actually expects blocking to work similar to centralized services: no replies can be made. posting followers-only is not a solution, either, because these people want their posts to be "public", but for differing definitions of "public".

**Comment by PleromaUser** [comment]
And again, they could just host their own website and reply there if they choose.  There's no way to make a block like you suggest, but not federating blocks stops an avenue of abuse *and* stops the gaslighting of blocked users.

**Comment by Gargron** [comment]
It's not gaslighting if it works in the majority of cases between cooperating servers. It's definitely not worse than straight out allowing other servers to make replies regardless of blocks in *all* cases.

**Comment by PleromaUser** [comment]
No, the behavior of "hiding blocker's post from blockee and trying to make the user act as if its deleted" *is gaslighting*.  Allowing other servers to make replies is harmless.

**Comment by trwnh** [comment]
>And again, they could just host their own website and reply there if they choose

And again, "of course it is always possible to host a web page "replying" to any other web page", but hosting your own website and making a disconnected reply is not the same, because most servers generally don't go looking for every web page that has a hyperlink to one of their posts in order to attach replies. The issue is about discovery and proliferation within the activitystreams/pub semweb.

Once again, in order to prevent a native reply, either you deliver a Block, or you require a token of validity (as discussed above). If it were harmless to allow replies, then no one would care. But they do, because it's not harmless. Being able to reply directly to someone that has blocked you is worse than being able to reply indirectly to someone that has blocked you.

**Comment by PleromaUser** [comment]
How is that harmful?

**Comment by Cassolotl** [comment]
Gaslighting is a sustained and deliberate effort to undermine someone's confidence in their own sanity so it doesn't really apply here. Withholding information isn't gaslighting.

**Comment by trwnh** [comment]
#11562 should be linked to this fwiw

**Comment by nemobis** [comment]
The original description of this issue is about two completely different things:

> Blocks are not sent to remote servers

and

> preventing blocked users from knowing they're blocked.

The second, under a literal interpretation of "preventing", is impossible in mainline Mastodon: if the blocking account is protected, the blocked user will need to unfollow; if it's public, the blocked user can simply compare public posts and what they receive. So this objective requires not having anything public. Locking down all accounts by default in an instance is a valid objective which is currently pursued mostly by the glitch and hometown forks; maybe this is a feature request for them? (Or maybe they already implemented it; I'm rarely able to find any public posts or accounts from hometown instances other than through my follows.)

So I'll focus on the title of the issue, "Blocks should not be federated". Not sending blocks to remote instances sounds to me like a request that the home server do all or most of the work that a block is expected to. I think there are three potential interpretations or levels to this:
1) Continue telling other instances about blocks, but don't expect the blocks to be honored; do something more on top for some of the things blocks are expected to do.
2) Stop telling other instances about blocks explicitly, but accept they may find out anyway, and do yourself all the things blocks are expected to do.
3) Same as 2) but also try to prevent hostile instances from knowing or guessing what's going on.

As for (1), it's relatively easy for some things. If the focus is on replies: the blocking user's instance could do some extra work to hide the replies which somehow still arrive from blocked users. (Including letting the admin know so that they can defederate the hostile instance?)

As for (2), I assume it would require to enable [AUTHORIZED_FETCH](https://docs.joinmastodon.org/admin/config/#authorized_fetch), so that more responses are tied to the identity of the user originating the request. To do *all* the things, some cooperation between instances is probably required and I'm not sure what can be done with the current ActivityPub protocol. For example, the blocking user's instance could just reject replies from blocked users; however, asking the blocked user's instance to also make them invisible for other instances would be another matter.

As for (3), I guess it would require deception on top of (2), similar to the way the GAFAM email providers break the email protocols and assume every mail relay out there is malicious until proven otherwise. Deceptive "anti-abuse" methods in the playbook include silently discarding things, rejecting things with deceptive or outright false error messages and so on.

**Comment by mjankowski** [comment]
This is pretty old issue ... is this sort of in an "we think we are following the spec and despite being aware of the tradeoffs involved we are in the least bad state" sort of scenario (and thus, can prob close as wontfix?) ... or is there more to resolve here?

**Comment by trwnh** [comment]
“more to do here” probably could include revisiting #11562 or at least making a decision/statement about that PR’s viability. it got closed with no real comment but i still think it’s worth doing.

aside from that, work at the AP Trust and Safety TF is scoped to include formalizing the concept of “moderation activities” which are not sent directly to the “user inbox” or to the public “shared inbox” but are instead sent to some “moderation inbox” or “moderations actor”. so it could be that in the future, we only send S2S Block activities if we detect the presence of a valid inbox for service-wide moderation contexts.

in any case the basic ask of “stop sending Block activities” is still worth considering due to the safety concerns with letting someone know they’ve been blocked. i know at least Pleroma makes it an admin config option whether to send Block activities or not.
